tests/sentry/event_manager/test_event_manager.py

```diff
@@ -2066,6 +2082,203 @@ def test_auto_update_grouping(self):
+            provider="integrations:github",
+            organization_id=self.project.organization.id,
+            integration_id=self.integration.id,
+            config={"name": self.repo_name},
+        )
+        RepositoryProjectPathConfig.objects.create(
+            repository_id=str(self.repo.id),
+            project_id=str(self.project.id),
+            organization_integration_id=str(self.org_integration.id),
+            stack_root="/stack/root",
+            source_root="/source/root",
+            default_branch="main",
+        )
+        stub_installation_token()
+        responses.add(
+            "GET",
+            f"https://api.github.com/repos/{self.repo_name}/commits/{LATER_COMMIT_SHA}",
+            json=json.loads(GET_COMMIT_EXAMPLE),
+        )
+        responses.add(
+            "GET",
+            f"https://api.github.com/repos/{self.repo_name}/commits/{EARLIER_COMMIT_SHA}",
+            json=json.loads(GET_PRIOR_COMMIT_EXAMPLE),
+        )
+        self.dummy_commit_sha = "a" * 40
+        responses.add(
+            responses.GET,
+            f"https://api.github.com/repos/{self.repo_name}/compare/{self.dummy_commit_sha}...{LATER_COMMIT_SHA}",
+            json=json.loads(COMPARE_COMMITS_EXAMPLE_WITH_INTERMEDIATE),
+        )
+        responses.add(
+            responses.GET,
+            f"https://api.github.com/repos/{self.repo_name}/commits?sha={LATER_COMMIT_SHA}",
+            json=json.loads(GET_LAST_2_COMMITS_EXAMPLE),
+        )
+
+    @mock.patch("sentry.integrations.github.client.get_jwt", return_value=b"jwt_token_1")
+    @responses.activate
+    def test_autoassign_commits_on_sha_release_version(self, get_jwt):
+        with self.feature("projects:auto-associate-commits-to-release"):
+
+            # Create a release
+            release = Release.objects.create(
+                organization_id=self.project.organization.id, version="abcabcabc"
+            )
+            release.add_project(self.project)
+            # Create a commit
+            commit = Commit.objects.create(
+                organization_id=self.project.organization.id,
+                repository_id=self.repo.id,
+                key=self.dummy_commit_sha,
+            )
+            # Make a release head commit
+            ReleaseHeadCommit.objects.create(
+                organization_id=self.project.organization.id,
+                repository_id=self.repo.id,
+                release=release,
+                commit=commit,
+            )
+            # Make a new release with SHA checksum
+            with self.tasks():
+                _ = self.make_release_event(LATER_COMMIT_SHA, self.project.id)
+
+            release2 = Release.objects.get(version=LATER_COMMIT_SHA)
+            commit_list = list(
+                Commit.objects.filter(releasecommit__release=release2).order_by(
+                    "releasecommit__order"
+                )
+            )
+
+            assert len(commit_list) == 2
```

Comment:
I would like to see a test that has commits in between (one is sufficient).
I'm thinking of a product that may have different releases on different deployments.

Deployments:
- Node A has commit X deployed
- Node B has commit Y deployed
- Node C has commit Z deployed

A produces an error and release 1 is created.
C produces an error and a release 2 is created (Y & Z should be the commits of it).
B produces an error and a release 3 is created (Y should be the commit of it).

This may sound weird but I'm pretty sure we will get into this situation.
We cannot assume a linear creation of releases.

Another situation I can think of is if developers branch of from their default branch and they do an X% release from that branch. The commit inclusion should take into consideration to which branch each release belongs to.

We may release this feature as-is behind a flag but we should not go public before we can handle this.

Reply:
API-2864 created to track this

Reply:
Thanks for filing it!